File: php-checks/src/main/java/org/sonar/php/checks/security/XPathUsageCheck.java
/*
 * SonarQube PHP Plugin
 * Copyright (C) 2010-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.php.checks.security;

import java.util.function.Predicate;
import org.sonar.check.Rule;
import org.sonar.php.checks.utils.type.FunctionCall;
import org.sonar.php.checks.utils.type.NewObjectCall;
import org.sonar.php.checks.utils.type.ObjectMemberFunctionCall;
import org.sonar.php.checks.utils.type.TreeValues;
import org.sonar.php.checks.utils.type.TypePredicateList;
import org.sonar.plugins.php.api.tree.Tree;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;

@Rule(key = "S4817")
public class XPathUsageCheck extends PHPVisitorCheck {

  private static final String MESSAGE = "Make sure that executing this XPATH expression is safe.";

  private static final Predicate<TreeValues> XPATH_PREDICATES = new TypePredicateList(
    new ObjectMemberFunctionCall("query", new NewObjectCall("DOMXpath")),
    new ObjectMemberFunctionCall("evaluate", new NewObjectCall("DOMXpath")),
    new ObjectMemberFunctionCall("xpath",
      new NewObjectCall("SimpleXMLElement"),
      new FunctionCall("simplexml_load_file"),
      new FunctionCall("simplexml_load_string"),
      new FunctionCall("simplexml_import_dom"))
    );

  @Override
  public void visitFunctionCall(FunctionCallTree tree) {
    TreeValues possibleValues = TreeValues.of(tree, context().symbolTable());
    if (XPATH_PREDICATES.test(possibleValues) && firstArgIsNotHardcoded(tree)) {
      context().newIssue(this, tree, MESSAGE);
    }
    super.visitFunctionCall(tree);
  }

  private static boolean firstArgIsNotHardcoded(FunctionCallTree tree) {
    return !tree.arguments().isEmpty() && !tree.arguments().get(0).is(Tree.Kind.REGULAR_STRING_LITERAL);
  }

}

File: php-checks/src/test/resources/checks/security/XPathUsageCheck.php
<?php

function evaluate_xpath($doc, $xpathstring, $xmlstring)
{
    $xpath = new DOMXpath($doc);
    $xpath->query($xpathstring); // Noncompliant {{Make sure that executing this XPATH expression is safe.}}
    $xpath->evaluate($xpathstring); // Noncompliant

    // There is no risk if the xpath is hardcoded
    $xpath->query("/users/user[@name='alice']"); // Compliant
    $xpath->evaluate("/users/user[@name='alice']"); // Compliant

    $xpath2 = NULL;
    if ($doc) {
        $xpath2 = new DOMXpath($doc);
    }
    $xpath2->query($xpathstring); // Noncompliant

    // An issue will also be created if the SimpleXMLElement is created
    // by simplexml_load_file, simplexml_load_string or simplexml_import_dom
    $xml = new SimpleXMLElement($doc);
    $xml->xpath($xpathstring); // Noncompliant

    // There is no risk if the xpath is hardcoded
    $xml->xpath("/users/user[@name='alice']"); // Compliant

    // simplexml_load_file returns a SimpleXMLElement
    $xml2 = simplexml_load_file('test.xml');
    $xml2->xpath($xpathstring); // Noncompliant
    // There is no risk if the xpath is hardcoded
    $xml2->xpath("/users/user[@name='alice']"); // Compliant

    // simplexml_load_string returns a SimpleXMLElement
    $xml3 = simplexml_load_string($xmlstring);
    $xml3->xpath($xpathstring); // Noncompliant
    // There is no risk if the xpath is hardcoded
    $xml3->xpath("/users/user[@name='alice']"); // Compliant

    // simplexml_import_dom returns a SimpleXMLElement
    $xml4 = simplexml_import_dom($xmlstring);
    $xml4->xpath($xpathstring); // Noncompliant
    // There is no risk if the xpath is hardcoded
    $xml4->xpath("/users/user[@name='alice']"); // Compliant

    // coverage
    $xpath->evaluate();
    $xpath->xpath($xpathstring);
    $xml->query($xpathstring);
    $xml->evaluate($xpathstring);

    $other = new Other($doc);
    $other->query($xpathstring);
    $other->evaluate($xpathstring);
    $other->xpath($xpathstring);
}
